chronos/asyncloop.nim

```diff
@@ -721,11 +748,47 @@ elif unixPlatform:
     # poll() call.
     loop.processCallbacks()
 
+    # Cleanup canceled timers
+    loop.removeTimers()
+
 else:
   proc initAPI() = discard
   proc globalInit() = discard
 
-proc addTimer*(at: Moment, cb: CallbackFunc, udata: pointer = nil) =
+proc setTimer*(at: Moment, cb: CallbackFunc, udata: pointer = nil): TimerHandle =
+  ## Arrange for the callback ``cb`` to be called at the given absolute
+  ## timestamp ``at``. You can also pass ``udata`` to callback.
+  ## Returns an opaque handle to be passed to ``clearTimer()`` to cancel
+  ## a pending timer.
+  let loop = getGlobalDispatcher()
+
+  var timer = TimerCallback(finishAt: at,
+                            function: AsyncCallback(function: cb, udata: udata))
+  loop.timerHandles.add(timer)
+  result = TimerHandle(loop.timerHandles.len - 1)
+
+  # try to optimize for the case where the top
+  # element is already removed and just replace it
+  if loop.timers.len > 0 and loop.timers[0].removed:
+    discard loop.timers.replace(timer)
+  else:
+    loop.timers.push(timer)
+
+proc clearTimer*(handle: TimerHandle) =
+  ## Clear the timer associated with the passed handle
+  var loop = getGlobalDispatcher()
+  var timer = loop.timerHandles[handle.Natural]
+  loop.timerHandles[handle.Natural] = nil
+
+  # optimize for handle being at the top
+  if timer == loop.timers[0]:
+    discard loop.timers.pop()
```

Comment:
What if the second best timer was also previously removed? Shouldn't this go into a loop here popping timers until it finds a one that's not removed?

Reply:
Not sure what you mean, this will run sequentially, so nothing should modify this.